src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,319 @@
+  // these facts around. This function is for generating the extra rule and
+  // it works similarly to "SpokenAttributeToDLIR".
+  DLIRAssertion SpokenCanActAsToDLIR(const Principal& speaker,
+                                     const CanActAs& can_act_as);
+
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIRInner(
+      const Principal& speaker, const Predicate& predicate) {
+    std::vector<DLIRAssertion> pred_list = {};
+    return std::make_pair(predicate, pred_list);
+  }
+
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIRInner(
+      const Principal& speaker, const Attribute& attribute) {
+    std::vector<DLIRAssertion> pred_list = {
+        SpokenAttributeToDLIR(speaker, attribute)};
+    return std::make_pair(AttributeToDLIR(attribute), pred_list);
+  }
+
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIRInner(
+      const Principal& speaker, const CanActAs& canActAs) {
+    std::vector<DLIRAssertion> pred_list = {
+        SpokenCanActAsToDLIR(speaker, canActAs)};
+    return std::make_pair(CanActAsToDLIR(canActAs), pred_list);
+  }
+
+  // The second return value represents 0 or 1 newly generated rules, so an
+  // option might seem more intuitive. However, the interface that consumes
+  // this needs to construct a vector anyway, so a vector is used in the
+  // return type.
+  std::pair<Predicate, std::vector<DLIRAssertion>> BaseFactToDLIR(
+      const Principal& speaker, const BaseFact& base_fact) {
+    return std::visit(
+        [this, &speaker](auto value) {
+          return BaseFactToDLIRInner(speaker, value);
+        },
+        base_fact.GetValue());
+  }
+
+  // This can result in 0 or more new rules because the translation of
+  // nested canSayFacts might result in more than 1 rule.
+  std::pair<Predicate, std::vector<DLIRAssertion>> FactToDLIR(
+      const Principal& speaker, const Fact& fact);
+
+  std::vector<DLIRAssertion> SingleSaysAssertionToDLIR(
+      const Principal& speaker, const Assertion& assertion);
+
+  std::vector<DLIRAssertion> SaysAssertionToDLIR(
+      const SaysAssertion& says_assertion) {
+    std::vector<DLIRAssertion> ret = {};
+    for (const Assertion& assertion : says_assertion.assertions()) {
+      std::vector<DLIRAssertion> single_translation =
+          SingleSaysAssertionToDLIR(says_assertion.principal(), assertion);
+      ret.insert(ret.end(), std::make_move_iterator(single_translation.begin()),
+                 std::make_move_iterator(single_translation.end()));
+    }
+    return ret;
+  }
+
+  std::vector<DLIRAssertion> SaysAssertionsToDLIR(
+      const std::vector<SaysAssertion>& says_assertions) {
+    std::vector<DLIRAssertion> ret = {};
+    for (const SaysAssertion& says_assertion : says_assertions) {
+      auto single_translation = SaysAssertionToDLIR(says_assertion);
+      ret.insert(ret.end(), std::make_move_iterator(single_translation.begin()),
+                 std::make_move_iterator(single_translation.end()));
```

Comment:
Can you make this into an utility function and use it here and other places in this PR? 

(See the `&&` for `src` in the `MoveAppend, it is not a type. :) )
```C++
template <typename T> 
void MoveAppend(std::vector<T>& dst, std::vector<T>&& src) {
  dst.insert(
    dst.end(), 
    std::make_move_iterator(src.begin()), 
    std::make_move_iterator(src.end()));
}
```

Then, we can use as follows here:
```C++
  auto single_translation = SaysAssertionToDLIR(says_assertion)
  MoveAppend(ret, std::move(single_translation));
```

Once this PR is in, we can refactor `MoveAppend` to `src/utils`.

Reply:
Oh nice. OK done.